src/bounce_solution/action_calculation.cpp

```diff
@@ -0,0 +1,1707 @@
+          // step, used to upgrade the step size.
+
+  double delta0; // Wanted precision
+  double delta1; // Step precision
+  int it;        // Counter
+
+  for (it = 0;
+       (it < maxiter) &&
+       (((dl_drho.back() > error) && ((l.back() - L) / L < error)) || it < 5);
+       it++) // Take at least 3 steps (due to dldrho < 0 due to numerical
+             // errors)
+  {
+    RK5_step({l.back(), dl_drho.back()},
+             {dl_drho.back(), d2l_drho2.back()},
+             2,
+             rho.back(),
+             step,
+             next_l_dldrho,
+             err);
+
+    delta1 = std::max(abs(err[0]), abs(err[1]));
+    delta0 = eps_abs * std::max(abs(l.back() + next_l_dldrho[0]),
+                                abs(dl_drho.back() + next_l_dldrho[1]));
+
+    // Update step list
+    rho.push_back(rho.back() + step);
+    l.push_back(next_l_dldrho[0]);
+    dl_drho.push_back(next_l_dldrho[1]);
+    d2l_drho2.push_back(d2ldrho2(l.back(), rho.back(), dl_drho.back()));
+
+    if (max_step > 0)
+    {
+      step = std::min((step * std::pow(delta0 / delta1, 0.2)), max_step);
+    }
+    else
+    {
+      step = step * std::pow(delta0 / delta1, 0.2) / 2;
+    }
+  }
+  if ((abs(dl_drho.back()) <= error) && (abs(l.back() - L) / L <= error))
+  {
+    ss << "Converged\t" << it << "\t" << l0 << "\t" << rho.back() << "\t"
+       << l.back() << "\t" << dl_drho.back();
+    conv = UndershootOvershootStatus::Converged;
+  }
+  else if (dl_drho.back() <= error)
+  {
+    ss << "Undershoot\t" << it << "\t" << l0 << "\t" << rho.back() << "\t"
+       << l.back() << "\t" << dl_drho.back();
+    rho.pop_back();
+    l.pop_back();
+    dl_drho.pop_back();
+    d2l_drho2.pop_back();
+    conv          = UndershootOvershootStatus::Undershoot;
+    UndershotOnce = true;
+  }
+  else if ((l.back() - L) / L >= error)
+  {
+    ss << "Overshoot\t" << it << "\t" << l0 << "\t" << rho.back() << "\t"
+       << l.back() << "\t" << dl_drho.back();
+    conv         = UndershootOvershootStatus::Overshoot;
+    OvershotOnce = true;
+  }
+  else
+  {
+    // This shouldnt happen
+    conv = UndershootOvershootStatus::Overshoot;
+  }
+  BSMPT::Logger::Write(BSMPT::LoggingLevel::BounceDetailed, ss.str());
+}
+
+double BounceActionInt::BackwardsPropagation()
+{
+  std::stringstream ss;
+  // Backwards propagation starting point finder
+  double l0  = 0;
+  double l00 = 1e100;
+  int i;
+  for (i = 0; i < 100; i++)
+  {
+    l00 = l0;
+    l0 -= Calc_dVdl(l0) / Calc_d2Vdl2(l0);
+    if (abs((l0 - l00) / Spline.L) < 1e-8 && Calc_dVdl(l0) >= 0)
+    {
+      if (l0 <= Spline.L / 100)
+      {
+        return l0;
+      }
+      ss << "Backwards propagation did not work...\t" << l0
+         << "\t using minus gradient method instead\n";
+      BSMPT::Logger::Write(BSMPT::LoggingLevel::BounceDetailed, ss.str());
+      ss.str(std::string());
+
+      l0  = 0;
+      l00 = 1e100;
+      for (i = 0; i < 1000; i++)
```

Comment:
Here you use the same index for the inner and outer loop. This seems like a mistake